### PR TITLE
support cpu/qemu only support SSE42

### PR DESCRIPTION
**Pull request by matrixji**
Issue: #12335

Signed-off-by: Ji Bin <matrixji@live.com>

### CODE REVIEW

**Comment by cydrain** [comment]
Hi Matrixji,

I have reviewed your PR, but I don't think it's a correct fix to let Milvus run on Apple M1 chip.

"-mf16c" is a gcc compile option, it's used to enable some pre-defined MACROs, same as "-mavx2".
We can use following command to see which MACROs have been enabled:

```
❯ gcc -mf16c -dM -E - < /dev/null | grep -E -i "sse|avx"
#define __SSE4_1__ 1
#define __SSE4_2__ 1
#define __SSE2_MATH__ 1
#define __AVX__ 1
#define __SSE_MATH__ 1
#define __SSE2__ 1
#define __SSSE3__ 1
#define __SSE__ 1
#define __SSE3__ 1

❯ gcc -mavx2 -dM -E - < /dev/null | grep -E -i "sse|avx"
#define __SSE4_1__ 1
#define __SSE4_2__ 1
#define __SSE2_MATH__ 1
#define __AVX__ 1
#define __AVX2__ 1
#define __SSE_MATH__ 1
#define __SSE2__ 1
#define __SSSE3__ 1
#define __SSE__ 1
#define __SSE3__ 1
```

"-mavx2" enables all MACROs enabled by "-mf16c" and "__AVX2__".

So, in your PR, it's redundant to use "-mf16c" and "-mavx2" together.

In the code under "faiss/utils" and "faiss/impl", 
all files with suffix "_avx" are for AVX2;
all files with suffix "_avx512" are for AVX512;
other files are for SSE4_2/AVX by default.

So it's reasonable to see _mm256_setr_epi8() in file distances_simd_avx.cpp.

**Comment by matrixji** [comment]
For the first change, yes, no need for  "-mf16c" and "-mavx2" together. however, before this change, they are used together already while compiling avx2 targets. previously "-mf16c" is passing from:
```
"CXXFLAGS=${EP_CXX_FLAGS} -mf16c -O3"
```
And `-mfc16` and is still needed for avx512 targets. avx512 targets need avx2 instructions by current code.


Calling _mm256_setr_epi8() in _avx.cpp is reasonable, but it shouldn't be in any initialization part, milvus detect supported instructions at runtime, so some unsupported instructions will be compiled but not executed at runtime. But initialization code like global variables will always be executed when the program startup.

```
const __m256i lookup = _mm256_setr_epi8(..)
```
this code will always be executed even before the main function.

**Comment by matrixji** [comment]
One more interesting to share, I'm trying to compile with avx/avx512 targets with below make rules(previously I have`-mf16c` and it succeed):
```makefile
%avx.o: %avx.cpp
        $(CXX) $(CPPFLAGS) $(CXXFLAGS) $(CPUFLAGS) -mavx2 -c $< -o $@

%avx512.o: %avx512.cpp
        $(CXX) $(CPPFLAGS) $(CXXFLAGS) $(CPUFLAGS) -mavx512f -mavx512dq -mavx512bw -c $< -o $@
```
But it fails both for avx objs and avx512 objs

Then I've checked the diff  between `-mf16c` and `-mavx2`, they have small diffs
```diff
--- output.mf16c.txt    2021-11-29 19:49:24.538348615 +0800
+++ output.mavx2.txt    2021-11-29 19:49:34.782657768 +0800
@@ -80,7 +80,6 @@
 #define __LDBL_MIN__ 3.36210314311209350626267781732175260e-4932L
 #define __STDC_UTF_16__ 1
 #define __DEC32_MAX__ 9.999999E96DF
-#define __F16C__ 1
 #define __FLT64X_DENORM_MIN__ 3.64519953188247460252840593361941982e-4951F64x
 #define __FLT32X_HAS_INFINITY__ 1
 #define __INT32_MAX__ 0x7fffffff
@@ -167,6 +166,7 @@
 #define __FLT128_MAX_EXP__ 16384
 #define __FLT32_MANT_DIG__ 24
 #define __FLOAT_WORD_ORDER__ __ORDER_LITTLE_ENDIAN__
+#define __AVX2__ 1
 #define __STDC_IEC_559_COMPLEX__ 1
 #define __FLT128_HAS_DENORM__ 1
 #define __FLT128_DIG__ 33
```

And I see inline functions in /usr/lib/gcc/x86_64-linux-gnu/7/include/f16cintrin.h (which included in ScalarQuantizerCodec_avx512.h and ScalarQuantizerCodec_avx.h) requires gcc target("f16c").

I'm trying to find out which Intel-based CPU started to support AVX2 and F16C, from intel's reference, AVX2 supported from SandyBridge (2011), and F16C is from IvyBridge (2012). That is to say, for the physical CPU, if it supports F16C, it should be also supporting avx2.

So, based on that, I think in faiss compilations:
- for avx2 objects we need -mf16c and -mavx2
- for avx512 objects we need -mf16c and -mavx512
- for generic objects, maybe -msse4.2 could be added for performance tunning, or -mavx if want tune with AVX(But Apple's M1 Rosetta is not supporting right now).

File: internal/core/src/index/thirdparty/CMakeLists.txt [hunk 1]
```diff
@@ -74,7 +74,7 @@ macro(build_faiss)
             "CXX=${FAISS_CXX_COMPILER}"
             "NVCC=${FAISS_CUDA_COMPILER}"
             "CFLAGS=${EP_C_FLAGS}"
-            "CXXFLAGS=${EP_CXX_FLAGS} -mf16c -O3"
+            "CXXFLAGS=${EP_CXX_FLAGS} -O3"
```
Comment (cydrain) [inline internal/core/src/index/thirdparty/CMakeLists.txt:77]: change to "CXXFLAGS=${EP_CXX_FLAGS} -msse4.2 -O3"

File: internal/core/src/index/thirdparty/faiss/utils/distances_simd_avx.cpp [hunk 2]
```diff
@@ -253,16 +253,16 @@ float fvec_Linf_avx (const float* x, const float* y, size_t d) {
...
-                /* 0 */ 0, /* 1 */ 1, /* 2 */ 1, /* 3 */ 2,
-                /* 4 */ 1, /* 5 */ 2, /* 6 */ 2, /* 7 */ 3,
-                /* 8 */ 1, /* 9 */ 2, /* a */ 2, /* b */ 3,
-                /* c */ 2, /* d */ 3, /* e */ 3, /* f */ 4,
-
-                /* 0 */ 0, /* 1 */ 1, /* 2 */ 1, /* 3 */ 2,
-                /* 4 */ 1, /* 5 */ 2, /* 6 */ 2, /* 7 */ 3,
-                /* 8 */ 1, /* 9 */ 2, /* a */ 2, /* b */ 3,
-                /* c */ 2, /* d */ 3, /* e */ 3, /* f */ 4
+#define DECLARE_LOOKUP \
```
Comment (cydrain) [inline internal/core/src/index/thirdparty/faiss/utils/distances_simd_avx.cpp:256]: don't change this file

**Comment by cydrain** [comment]
Hi Matrixji,

I get your point.

I give one comment, please have a check.

**Comment by cydrain** [comment]
please update your code, I'll approve.

**Comment by matrixji** [comment]
> please update your code, I'll approve.

Thanks for your comments, I'll try some testing with SSE options to check which SSE option is Maximum supported by Rosetta.

**Comment by matrixji** [comment]
Tested with -msse4.2 with docker desktop under macOS (M1). So update the generic flags with -msse4.2.

**Comment by scsven** [comment]
Hi @matrixji, thanks for your contributions. But the commit message should start with a capital letter, so that please reword the commit message. 

More rules about code review can be found here: https://github.com/milvus-io/milvus/blob/master/CODE_REVIEW.md#for-approvers

**Pull request merged**